File: javascript/largest-series-product/largest-series-product.js
export const largestProduct = (digitSequence, length) => {
	if (length < 0) { throw new Error('Span must be greater than zero'); }
	if (!/^\d*$/.test(digitSequence)) { throw new Error('Digits input must only contain digits'); }

	let sumOfList = (items) => items.reduce((accumulator, currentValue) => accumulator * currentValue);

	if (digitSequence.length < length) { throw new Error('Span must be smaller than string length'); }
	if (length === 0) { return 1; }

	let chunkOfProducts = groupOfAllProducts(digitSequence, length);

	return findLargestProductResultInGroup(chunkOfProducts, sumOfList);
};


function groupOfAllProducts(digitSequence, length) {
	let chunkOfProducts = [];

	for (let i = 0; i <= digitSequence.length - length; i++) {
		chunkOfProducts.push([...digitSequence].map(Number).slice(i, i + length));
	}

	return chunkOfProducts;
}


function findLargestProductResultInGroup(chunkOfProducts, sumOfList) {
	let largestProduct = null;

	for (let i = 0; i < chunkOfProducts.length; i++) {
		if (largestProduct === null) {
			largestProduct = sumOfList(chunkOfProducts[i]);
		}
		else {
			let currentProductEvaluation = sumOfList(chunkOfProducts[i]);

			if (largestProduct < currentProductEvaluation) {
				largestProduct = currentProductEvaluation;
			}
		}
	}

	return largestProduct;
}

File: python/largest-series-product/largest_series_product.py
from functools import reduce
from operator import mul

def largest_product(series: str, size: int) -> int:
    if size < 0:
        raise ValueError("span must not be negative")
    
    series_length = len(series)

    if series_length < size:
        raise ValueError("span must be smaller than string length")
    
    if series_length == 0 or size == 0:
        return 1

    if not series.isdigit():
        raise ValueError("digits input must only contain digits");
 
    return  max(([reduce(mul, map(int, digits)) 
                  for digits in sliding_window(series, size)]))


def sliding_window(seq: str, size: int) -> list[str]:
    for start_index in range(len(seq) - size + 1):
        yield seq[start_index: start_index + size] 